File: prayer-sdk-ts/src/actions.ts
import type { Action } from "./generated/types.js";
import type { GoTarget } from "./conveniences.js";
export const undock = (): Action => ({ type: "undock" });
export const dock = (): Action => ({ type: "dock" });
export const wait = (ticks: number): Action => ({ type: "wait", request: { ticks } });
export const go = (target: { poi: string } | { system: string } | GoTarget): Action => {
  const destination: GoTarget = "poi" in target
    ? { kind: "poi", value: target.poi }
    : "system" in target
      ? { kind: "system", value: target.system }
      : target;
  return { type: "go", request: { destination } };
};

type ActionType = Action["type"];
type RequestFor<T extends ActionType> = Action extends infer A
  ? A extends { type: infer K; request: infer R }
    ? T extends K ? R : never
    : never
  : never;
type RequestActionType = { [T in ActionType]: [RequestFor<T>] extends [never] ? never : T }[ActionType];
type UnitActionType = Exclude<ActionType, RequestActionType>;
type FullyNullableRequestActionType = "mine" | "scan" | "repair" | "refuel" | "distress_signal";

/**
 * Canonical empty requests for commands whose payload fields are all nullable.
 * The mapped type makes this registry exhaustive: adding or changing one of
 * these requests in the generated Action union fails compilation here.
 */
const nullableRequestDefaults = {
  mine: { resource: null },
  scan: { target: null },
  repair: { target: null, quantity: null, item: null },
  refuel: { target: null, quantity: null, item: null },
  distress_signal: { distress_type: null },
} satisfies Record<FullyNullableRequestActionType, Record<string, unknown>>;

const nullableRequested = <T extends FullyNullableRequestActionType>(type: T) =>
  (request: Partial<RequestFor<T>> = {}): Action => {
    const normalized: Record<string, unknown> = { ...nullableRequestDefaults[type] };
    for (const [field, value] of Object.entries(request)) {
      if (value !== undefined) normalized[field] = value;
    }
    return action(type, normalized as RequestFor<T>);
  };

export const mine = (resource?: string): Action => action("mine", { resource: resource ?? null });

/** Build any canonical prayer-actions operation with its exact request type. */
export function action<T extends RequestActionType>(type: T, request: RequestFor<T>): Action;
export function action<T extends UnitActionType>(type: T): Action;
export function action(type: ActionType, request?: unknown): Action {
  return (request === undefined ? { type } : { type, request }) as Action;
}

const requested = <T extends RequestActionType>(type: T) => (request: RequestFor<T>): Action => action(type, request);
const unit = <T extends UnitActionType>(type: T) => (): Action => action(type);

/** Complete helper catalog for the canonical prayer-actions wire shape. */
export const actions = {
  undock, dock, wait, mine, go,
  halt: unit("halt"),
  transfer: requested("transfer"), setHome: unit("set_home"), find: requested("find"), survey: unit("survey"),
  attack: requested("attack"), scan: nullableRequested("scan"), cloak: requested("cloak"), hunt: requested("hunt"), prepayTax: requested("prepay_tax"),
  acceptMission: requested("accept_mission"), abandonMission: requested("abandon_mission"), declineMission: requested("decline_mission"), completeMission: requested("complete_mission"),
  loadPassenger: requested("load_passenger"), unloadPassenger: requested("unload_passenger"), buy: requested("buy"), sell: requested("sell"),
  cancelBuy: requested("cancel_buy"), cancelSell: requested("cancel_sell"), factionCreate: requested("faction_create"), factionInvite: requested("faction_invite"),
  factionAcceptInvite: requested("faction_accept_invite"), factionKick: requested("faction_kick"), factionSetRole: requested("faction_set_role"),
  facilityBuild: requested("facility_build"), factionFacilityBuild: requested("faction_facility_build"), facilityUpgrade: requested("facility_upgrade"),
  factionFacilityUpgrade: requested("faction_facility_upgrade"), facilityDismantle: requested("facility_dismantle"), factionFacilityDismantle: requested("faction_facility_dismantle"),
  facilitySetAccess: requested("facility_set_access"), facilitySetOutputPrice: requested("facility_set_output_price"), facilitySetName: requested("facility_set_name"),
  useItem: requested("use_item"), repair: nullableRequested("repair"), repairModule: requested("repair_module"), recycle: requested("recycle"), refuel: nullableRequested("refuel"),
  selfDestruct: unit("self_destruct"), switchShip: requested("switch_ship"), renameShip: requested("rename_ship"), installMod: requested("install_mod"), uninstallMod: requested("uninstall_mod"),
  buyShip: requested("buy_ship"), buyListedShip: requested("buy_listed_ship"), commissionShip: requested("commission_ship"), sellShip: requested("sell_ship"), scrapShip: requested("scrap_ship"),
  listShipForSale: requested("list_ship_for_sale"), refitShip: unit("refit_ship"), cancelCommission: requested("cancel_commission"), supplyCommission: requested("supply_commission"),
  cancelShipListing: requested("cancel_ship_listing"), placeShipBuyOrder: requested("place_ship_buy_order"), cancelShipBuyOrder: requested("cancel_ship_buy_order"),
  sellShipToOrder: requested("sell_ship_to_order"), cancelOrder: requested("cancel_order"), modifyOrder: requested("modify_order"), craft: requested("craft"), cancelCraftJob: requested("cancel_craft_job"),
  salvageWreck: requested("salvage_wreck"), towWreck: requested("tow_wreck"), scrapWreck: unit("scrap_wreck"), sellWreck: unit("sell_wreck"), releaseWreck: unit("release_wreck"), insureShip: requested("insure_ship"),
  citizenshipApply: requested("citizenship_apply"), citizenshipWithdraw: requested("citizenship_withdraw"), citizenshipRenounce: requested("citizenship_renounce"),
  tradeOffer: requested("trade_offer"), tradeAccept: requested("trade_accept"), factionLeave: unit("faction_leave"), factionWithdrawInvite: requested("faction_withdraw_invite"),
  factionProposeAlly: requested("faction_propose_ally"), factionAcceptAlly: requested("faction_accept_ally"), factionRemoveAlly: requested("faction_remove_ally"),
  factionDeclareWar: requested("faction_declare_war"), factionProposePeace: requested("faction_propose_peace"), factionAcceptPeace: requested("faction_accept_peace"),
  factionSetEnemy: requested("faction_set_enemy"), factionRemoveEnemy: requested("faction_remove_enemy"), factionPrepayTax: requested("faction_prepay_tax"),
  factionCancelMission: requested("faction_cancel_mission"), espionage: unit("espionage"), scanPoi: requested("scan_poi"), distressSignal: nullableRequested("distress_signal"), say: requested("say"),
} as const;

File: prayer-sdk-py/src/prayer_sdk/actions.py
"""Exact-wire action constructors."""

from __future__ import annotations

from collections.abc import Mapping
from types import MappingProxyType
from typing import Any

Action = Mapping[str, Any]


def action(type: str, request: Mapping[str, Any] | None = None, *, include_request: bool | None = None) -> Action:
    value: dict[str, Any] = {"type": type}
    if request is not None or include_request:
        value["request"] = MappingProxyType(dict(request or {}))
    return MappingProxyType(value)


def undock() -> Action: return action("undock")
def dock() -> Action: return action("dock")
def wait(ticks: int) -> Action: return action("wait", {"ticks": ticks})
def mine(resource: str | None = None) -> Action: return action("mine", {"resource": resource})
def go(*, poi: str | None = None, system: str | None = None,
       kind: str | None = None, value: str | None = None) -> Action:
    choices = sum(x is not None for x in (poi, system, kind))
    if choices != 1:
        raise ValueError("go requires exactly one of poi, system, or kind/value")
    destination = {"kind": "poi", "value": poi} if poi is not None else (
        {"kind": "system", "value": system} if system is not None else {"kind": kind, "value": value}
    )
    return action("go", {"destination": destination})


_UNIT = {"halt", "set_home", "survey", "self_destruct", "refit_ship", "scrap_wreck",
         "sell_wreck", "release_wreck", "faction_leave", "espionage"}
_NULLABLE = {
    "scan": {"target": None}, "repair": {"target": None, "quantity": None, "item": None},
    "refuel": {"target": None, "quantity": None, "item": None},
    "distress_signal": {"distress_type": None},
}


def _named_helper(type: str):
    if type in _UNIT:
        return lambda: action(type)
    if type in _NULLABLE:
        return lambda **request: action(type, {**_NULLABLE[type], **request})
    return lambda request=None, **fields: action(type, request if request is not None else fields)


ACTION_TYPES = (
    "undock dock wait mine go halt transfer set_home find survey attack scan cloak hunt prepay_tax "
    "accept_mission abandon_mission decline_mission complete_mission load_passenger unload_passenger buy sell "
    "cancel_buy cancel_sell faction_create faction_invite faction_accept_invite faction_kick faction_set_role found_station "
    "facility_build faction_facility_build facility_upgrade faction_facility_upgrade facility_dismantle "
    "faction_facility_dismantle facility_set_access facility_set_output_price facility_set_name use_item repair "
    "repair_module recycle refuel self_destruct switch_ship rename_ship install_mod uninstall_mod buy_ship "
    "buy_listed_ship commission_ship sell_ship scrap_ship list_ship_for_sale refit_ship cancel_commission "
    "supply_commission cancel_ship_listing place_ship_buy_order cancel_ship_buy_order sell_ship_to_order "
    "cancel_order modify_order craft cancel_craft_job salvage_wreck tow_wreck scrap_wreck sell_wreck release_wreck "
    "insure_ship citizenship_apply citizenship_withdraw citizenship_renounce trade_offer trade_accept faction_leave "
    "faction_withdraw_invite faction_propose_ally faction_accept_ally faction_remove_ally faction_declare_war "
    "faction_propose_peace faction_accept_peace faction_set_enemy faction_remove_enemy faction_prepay_tax "
    "faction_cancel_mission espionage scan_poi distress_signal say"
).split()

for _type in ACTION_TYPES:
    if _type not in {"undock", "dock", "wait", "mine", "go"}:
        globals()[_type] = _named_helper(_type)

actions = MappingProxyType({name: globals()[name] for name in ACTION_TYPES})
__all__ = ["ACTION_TYPES", "Action", "action", "actions", *ACTION_TYPES]
